go/pkg/pwapi/api_agent-update-state.go

```diff
@@ -15,17 +16,35 @@ func (svc *service) AgentUpdateState(ctx context.Context, in *AgentUpdateState_I
 		return nil, errcode.ErrMissingInput
 	}
 
-	for _, challengeInstance := range in.Instances {
-		cpy := challengeInstance
-		err := svc.db.Model(&cpy).
-			Update(pwdb.ChallengeInstance{
-				Status:         challengeInstance.Status,
-				InstanceConfig: challengeInstance.InstanceConfig,
-			}).
-			Error
-		if err != nil {
-			return nil, errcode.ErrAgentUpdateState.Wrap(err)
+	userID, err := userIDFromContext(ctx, svc.db)
+	if err != nil {
+		return nil, errcode.ErrGetUserIDFromContext.Wrap(err)
+	}
+
+	err = svc.db.Transaction(func(tx *gorm.DB) error {
+		for _, challengeInstance := range in.Instances {
+			cpy := challengeInstance
+			err := tx.Model(&cpy).
+				Update(pwdb.ChallengeInstance{
+					Status:         challengeInstance.Status,
+					InstanceConfig: challengeInstance.InstanceConfig,
+				}).
+				Error
+			if err != nil {
+				return errcode.ErrAgentUpdateState.Wrap(err)
+			}
+		}
```

Comment:
the current function updates everything, which is not a problem, but the activity should not be called for each existing instance but only for the ones that actually changed, so you need to add a new if condition to only create the activity if the db update changed something